File: app/src/main/java/com/example/readative/room/entity/Book.java
package com.example.readative.room.entity;

import androidx.annotation.NonNull;
import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.Index;
import androidx.room.PrimaryKey;

import com.example.readative.util.Converters;

import java.util.Date;

@Entity(indices = {@Index(value = {"book_uuid"}, unique = true)})
public class Book {
    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "book_id")
    public long id;
    @ColumnInfo(name = "book_uuid")
    public String uuid;

    @ColumnInfo(name = "author_owner_id")
    public int authorOwnerId;

    @NonNull
    @ColumnInfo(name = "book_title")
    public String title;
    @ColumnInfo(name = "book_description")
    public String description;
    @NonNull
    @ColumnInfo(name = "book_cover")
    public String coverUrl;
    @NonNull
    @ColumnInfo(name = "book_progress")
    public int progress;
    @NonNull
    @ColumnInfo(name = "book_status")
    public Converters.BookStatus status;
    @NonNull
    @ColumnInfo(name = "book_date_modified")
    public Date modified;
    @NonNull
    @ColumnInfo(name = "book_date_added")
    public Date added;
    @ColumnInfo(name = "book_date_opened")
    public Date lastOpened;
}

File: app/src/main/java/com/example/readative/persistance/entities/Book.kt
package com.example.readative.persistance.entities

import androidx.annotation.NonNull
import androidx.room.*
import androidx.room.ForeignKey.CASCADE
import androidx.room.ForeignKey.SET_NULL
import java.time.OffsetDateTime

@Entity(
    tableName = "books",
    indices = [
        Index(value = ["checksum"], unique = true),
        Index(value = ["author_id"])
    ],
    foreignKeys = [
        ForeignKey(
            entity = Author::class,
            parentColumns = ["id"],
            childColumns = ["author_id"],
            onUpdate = CASCADE,
            onDelete = SET_NULL
        )
    ]
)
data class Book (
    @PrimaryKey(autoGenerate = true) @ColumnInfo(name = "id") override val id: Long,
    @ColumnInfo(name = "author_id") val authorId: Long? = null,
    @ColumnInfo(name = "checksum") val checksum: String,
    @ColumnInfo(name = "title") val title: String,
    @ColumnInfo(name = "annotation") val description: String? = null,
    @ColumnInfo(name = "book_cover") val coverPath: String,
    @ColumnInfo(name = "reading_status") val status: ReadingStatus = ReadingStatus.UNREAD,
    @ColumnInfo(name = "reading_progress") val progress: Float = 0f,
    @ColumnInfo(name = "published_date") val published: OffsetDateTime,
    @ColumnInfo(name = "added_date") val added: OffsetDateTime? = null,
    @ColumnInfo(name = "last_opened") val opened: OffsetDateTime? = null
) : ReadativeEntity